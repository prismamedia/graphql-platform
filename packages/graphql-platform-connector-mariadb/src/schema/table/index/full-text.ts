import { Memoize } from '@prismamedia/ts-memoize';
import inflection from 'inflection';
import { escapeIdentifier } from '../../../escaping.js';
import type { LeafColumn } from '../../table.js';
import { AbstractIndex } from '../abstract-index.js';

/**
 * @see https://mariadb.com/kb/en/full-text-index-overview/
 */
export class FullTextIndex extends AbstractIndex {
  public readonly name: string;

  public constructor(public readonly column: LeafColumn) {
    super(column.table);

    this.name = `txt_${inflection.underscore(column.name)}`
      // @see https://mariadb.com/kb/en/identifier-names/#maximum-length
      .substring(0, 64);
  }

  /**
   * @see https://mariadb.com/kb/en/create-table/#unique
   */
  @Memoize()
  public get definition(): string {
    return `FULLTEXT ${escapeIdentifier(this.name)} (${escapeIdentifier(
      this.column.name,
    )})`;
  }
}
