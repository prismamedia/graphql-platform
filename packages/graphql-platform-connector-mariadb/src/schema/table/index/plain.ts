import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import { escapeIdentifier } from '../../../escaping.js';
import type { Table } from '../../table.js';
import { AbstractIndex } from '../abstract-index.js';
import type { Column } from '../column.js';

export interface PlainIndexConfig {
  /**
   * Optional, the index's name
   */
  name?: utils.Nillable<string>;

  /**
   * Required, the index's columns' name
   */
  columns: ReadonlyArray<Column['name']>;
}

/**
 * @see https://mariadb.com/kb/en/getting-started-with-indexes/#plain-indexes
 */
export class PlainIndex extends AbstractIndex {
  public constructor(
    table: Table,
    public readonly config: PlainIndexConfig,
    public readonly configPath: utils.Path,
  ) {
    utils.assertPlainObjectConfig(config, configPath);

    super(table);
  }

  @Memoize()
  public get columns(): ReadonlyArray<Column> {
    const config = this.config.columns;
    const configPath = utils.addPath(this.configPath, 'columns');

    return Object.freeze(
      config.map((columnName, index) => {
        const column = this.table.columns.find(
          (column) => column.name === columnName,
        );

        if (!column) {
          throw new utils.UnexpectedConfigError(
            'an existing column',
            columnName,
            { path: utils.addPath(configPath, index) },
          );
        }

        return column;
      }),
    );
  }

  @Memoize()
  public get name(): string {
    const config = this.config?.name;
    const configPath = utils.addPath(this.configPath, 'name');

    if (config) {
      if (typeof config !== 'string') {
        throw new utils.UnexpectedConfigError('a string', config, {
          path: configPath,
        });
      }

      if (config.length > 64) {
        throw new utils.UnexpectedConfigError(
          'an identifier shorter than 64 characters',
          config,
          { path: configPath },
        );
      }

      return config;
    } else if (this.table.schema.config?.naming?.plainIndex) {
      return this.table.schema.config.naming.plainIndex(
        this.table.name,
        this.columns,
      );
    }

    return (
      ['idx', ...this.columns.map(({ name }) => name)]
        .join('_')
        // @see https://mariadb.com/kb/en/identifier-names/#maximum-length
        .substring(0, 64)
    );
  }

  /**
   * @see https://mariadb.com/kb/en/create-table/#plain-indexes
   */
  @Memoize()
  public get definition(): string {
    return `INDEX ${escapeIdentifier(this.name)} (${this.columns
      .map(({ name }) => escapeIdentifier(name))
      .join(',')})`;
  }
}
