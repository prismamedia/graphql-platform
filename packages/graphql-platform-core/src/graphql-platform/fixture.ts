import { Entries, fromEntries } from '@prismamedia/graphql-platform-utils';
import { DepGraph } from 'dependency-graph';
import { OperationDefinitionNode, print } from 'graphql';
import { Logger } from 'winston';
import { GraphQLPlatform, GraphQLRequest } from '../graphql-platform';
import { CreateOneDataInputValue } from './operation';
import { AnyResource, Relation, SerializedFieldValue } from './resource';
import { TypeKind, WhereUniqueInputValue } from './type';

export class FixtureGraph extends DepGraph<Fixture> {}

export type FixtureFieldData = SerializedFieldValue;

export type FixtureRelationData = null | Fixture['name'];

export type FixtureComponentData = FixtureFieldData | FixtureRelationData;

export type FixtureData = {
  [componentName: string]: FixtureComponentData;
};

export type FixtureDataMap = {
  [fixtureName: string]: FixtureData;
};

export class Fixture {
  readonly logger?: Logger;
  protected id?: WhereUniqueInputValue;

  public constructor(
    readonly name: string,
    readonly resource: AnyResource,
    readonly data: FixtureData,
    protected graph: FixtureGraph,
    protected gp: GraphQLPlatform,
  ) {
    this.logger = gp.getLogger();
  }

  public toString(): string {
    return `${this.resource}.${this.name}`;
  }

  public getRelatedFixture(relation: Relation): Fixture | null {
    this.resource.getRelationMap().assert(relation);

    const relationValue = this.data[relation.name];

    if (relationValue == null) {
      if (relation.isRequired()) {
        throw new Error(`In the fixture "${this.name}", the relation "${relation}" can't be null.`);
      }
    } else if (!(typeof relationValue === 'string' && this.graph.hasNode(relationValue))) {
      throw new Error(
        `In the fixture "${this.name}", the relation "${relation}" has to reference an "${relation.getTo()}" fixture.`,
      );
    }

    return relationValue ? this.graph.getNodeData(relationValue) : null;
  }

  public getCreateMutationAST(): OperationDefinitionNode {
    const operationDefinitionNode: OperationDefinitionNode = {
      kind: 'OperationDefinition',
      operation: 'mutation',
      variableDefinitions: [
        {
          kind: 'VariableDefinition',
          variable: {
            kind: 'Variable',
            name: {
              kind: 'Name',
              value: 'data',
            },
          },
          type: {
            kind: 'NonNullType',
            type: {
              kind: 'NamedType',
              name: {
                kind: 'Name',
                value: this.resource.getMutation('CreateOne').getDataType().name,
              },
            },
          },
        },
      ],
      selectionSet: {
        kind: 'SelectionSet',
        selections: [
          {
            kind: 'Field',
            name: {
              kind: 'Name',
              value: this.resource.getMutation('CreateOne').name,
            },
            alias: {
              kind: 'Name',
              value: 'id',
            },
            arguments: [
              {
                kind: 'Argument',
                name: {
                  kind: 'Name',
                  value: 'data',
                },
                value: {
                  kind: 'Variable',
                  name: {
                    kind: 'Name',
                    value: 'data',
                  },
                },
              },
            ],
            selectionSet: this.resource
              .getFirstPublicUnique()
              .getSelectionNode(TypeKind.Output)
              .toAST(),
          },
        ],
      },
    };

    return operationDefinitionNode;
  }

  public getCreateMutationSource(): string {
    return print(this.getCreateMutationAST());
  }

  public getCreateMutationVariables(): GraphQLRequest['variableValues'] {
    return {
      data: fromEntries([
        // Fields
        ...[...this.resource.getFieldSet()].map(field =>
          !field.isFullyManaged() ? [field.name, this.data[field.name] as FixtureFieldData] : undefined,
        ),
        // Relations
        ...[...this.resource.getRelationSet()].map(relation => {
          if (!relation.isFullyManaged()) {
            const relatedFixture = this.getRelatedFixture(relation);
            if (relatedFixture) {
              return [relation.name, { connect: relatedFixture.assertId() }];
            }
          }
        }),
      ] as Entries<CreateOneDataInputValue>),
    };
  }

  public getCreateMutation(): GraphQLRequest {
    return {
      source: this.getCreateMutationSource(),
      variableValues: this.getCreateMutationVariables(),
    };
  }

  public async load(contextValue?: GraphQLRequest['contextValue']): Promise<void> {
    try {
      const { id } = await this.gp.execute<{ id: WhereUniqueInputValue }>({
        ...this.getCreateMutation(),
        contextValue,
      });

      this.id = this.resource.getInputType('WhereUnique').assert(id);
    } catch (error) {
      this.logger && this.logger.error(error);

      throw error;
    }
  }

  public assertId(): WhereUniqueInputValue {
    if (!this.id) {
      throw new Error(`The fixture "${this.name}" is not loaded yet.`);
    }

    return this.id;
  }
}
