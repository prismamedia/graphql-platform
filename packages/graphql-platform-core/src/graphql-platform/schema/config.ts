import { GraphQLOperationType } from '@prismamedia/graphql-platform-utils';
import { GraphQLFieldConfigMap, GraphQLObjectType, GraphQLSchemaConfig as BaseGraphQLSchemaConfig } from 'graphql';
import { OperationType } from '../operation';
import { ResourceMap } from '../resource/map';

export interface GraphQLSchemaConfigConfig
  extends Omit<BaseGraphQLSchemaConfig, 'mutation' | 'query' | 'subscription'> {
  resourceMap: ResourceMap;
  queryMap: GraphQLFieldConfigMap<any, any>;
  mutationMap: GraphQLFieldConfigMap<any, any>;
}

export class GraphQLSchemaConfig implements BaseGraphQLSchemaConfig {
  public constructor(readonly config: GraphQLSchemaConfigConfig) {}

  public getOperationTypeFieldMap<TType extends OperationType>(type: TType): GraphQLFieldConfigMap<any, any> {
    const fields: GraphQLFieldConfigMap<any, any> = {};

    for (const resource of this.config.resourceMap.values()) {
      for (const operation of resource.getOperations(type)) {
        if (operation.isPublic()) {
          fields[operation.name] = operation.getGraphQLFieldConfig();
        }
      }
    }

    return fields;
  }

  public get mutation() {
    return new GraphQLObjectType({
      name: 'Mutation',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Mutation), ...this.config.mutationMap },
    });
  }

  public get query() {
    return new GraphQLObjectType({
      name: 'Query',
      fields: { ...this.getOperationTypeFieldMap(GraphQLOperationType.Query), ...this.config.queryMap },
    });
  }

  public get subscription() {
    return undefined;
  }
}
