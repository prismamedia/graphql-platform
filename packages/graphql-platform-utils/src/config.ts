export * from './config/name.js';
export * from './config/optional-deprecation.js';
export * from './config/optional-description.js';
export * from './config/optional-flag.js';
