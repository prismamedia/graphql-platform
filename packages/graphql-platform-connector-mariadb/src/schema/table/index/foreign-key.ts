import * as core from '@prismamedia/graphql-platform';
import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/memoize';
import { escapeIdentifier } from '../../../escaping.js';
import type { MariaDBConnector } from '../../../index.js';
import type { Table } from '../../table.js';
import { AbstractIndex } from '../abstract-index.js';
import { ReferenceColumn } from '../column/reference.js';

export interface ForeignKeyIndexConfig {
  /**
   * Optional, the index's name
   */
  name?: utils.Nillable<string>;
}

/**
 * @see https://mariadb.com/kb/en/foreign-keys/
 */
export class ForeignKeyIndex extends AbstractIndex {
  public readonly config?: ForeignKeyIndexConfig;
  public readonly configPath: utils.Path;

  public constructor(
    table: Table,
    public readonly edge: core.Edge<any, MariaDBConnector>,
  ) {
    super(table);

    // config
    {
      this.config = edge.config.foreignKey;
      this.configPath = utils.addPath(edge.configPath, 'foreignKey');

      utils.assertNillablePlainObjectConfig(this.config, this.configPath);
    }
  }

  @Memoize()
  public get referencedTable(): Table {
    return this.table.schema.getTableByNode(this.edge.head);
  }

  @Memoize()
  public get columns(): ReadonlyArray<ReferenceColumn> {
    return this.table.getColumnTreeByEdge(this.edge).columns;
  }

  @Memoize()
  public get name(): string {
    const nameConfig = this.config?.name;
    const nameConfigPath = utils.addPath(this.configPath, 'name');

    if (nameConfig) {
      if (typeof nameConfig !== 'string') {
        throw new utils.UnexpectedConfigError('a string', nameConfig, {
          path: nameConfigPath,
        });
      }

      if (nameConfig.length > 64) {
        throw new utils.UnexpectedConfigError(
          'an identifier shorter than 64 characters',
          nameConfig,
          { path: nameConfigPath },
        );
      }

      return nameConfig;
    }

    return this.table.schema.namingStrategy.getForeignKeyIndexName(
      this.table.name,
      this.edge,
      this.columns,
    );
  }

  /**
   * @see https://mariadb.com/kb/en/create-table/#foreign-key
   */
  @Memoize()
  public get definition(): string {
    return [
      'FOREIGN KEY',
      escapeIdentifier(this.name),
      `(${this.columns.map(({ name }) => escapeIdentifier(name)).join(',')})`,
      'REFERENCES',
      escapeIdentifier(this.referencedTable.name),
      `(${this.columns
        .map(({ referencedColumn: { name } }) => escapeIdentifier(name))
        .join(',')})`,
      'ON UPDATE RESTRICT',
      'ON DELETE RESTRICT',
    ]
      .filter(Boolean)
      .join(' ');
  }
}
