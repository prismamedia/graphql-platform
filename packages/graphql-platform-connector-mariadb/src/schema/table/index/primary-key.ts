import * as core from '@prismamedia/graphql-platform';
import { Memoize } from '@prismamedia/memoize';
import assert from 'node:assert/strict';
import { escapeIdentifier } from '../../../escaping.js';
import type { MariaDBConnector } from '../../../index.js';
import type { Column, Table } from '../../table.js';
import { AbstractIndex } from '../abstract-index.js';

/**
 * @see https://mariadb.com/kb/en/getting-started-with-indexes/#primary-key
 */
export class PrimaryKey extends AbstractIndex {
  public readonly name: string;

  public constructor(
    table: Table,
    public readonly uniqueConstraint: core.UniqueConstraint<
      any,
      MariaDBConnector
    >,
  ) {
    assert(
      uniqueConstraint.isIdentifier(),
      `The "${uniqueConstraint}" unique-constraint is not the identifier`,
    );

    super(table);

    // name
    {
      this.name = 'PRIMARY KEY';
    }
  }

  @Memoize()
  public get columns(): ReadonlyArray<Column> {
    return Object.freeze(
      this.table.getColumnsByComponents(...this.uniqueConstraint.components),
    );
  }

  /**
   * @see https://mariadb.com/kb/en/create-table/#primary-key
   */
  @Memoize()
  public get definition(): string {
    return `PRIMARY KEY (${this.columns
      .map(({ name }) => escapeIdentifier(name))
      .join(',')})`;
  }
}
