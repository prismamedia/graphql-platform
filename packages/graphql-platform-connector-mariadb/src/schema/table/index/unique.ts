import * as core from '@prismamedia/graphql-platform';
import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/memoize';

import assert from 'node:assert/strict';
import { escapeIdentifier } from '../../../escaping.js';
import type { MariaDBConnector } from '../../../index.js';
import { ensureIdentifierName } from '../../naming-strategy.js';
import type { Column, Table } from '../../table.js';
import { AbstractIndex } from '../abstract-index.js';

export * from './unique/diagnosis.js';

export interface UniqueIndexConfig {
  /**
   * Optional, the index's name
   */
  name?: utils.Nillable<string>;
}

/**
 * @see https://mariadb.com/kb/en/getting-started-with-indexes/#unique-index
 */
export class UniqueIndex extends AbstractIndex {
  public readonly config?: UniqueIndexConfig;
  public readonly configPath: utils.Path;

  public constructor(
    table: Table,
    public readonly uniqueConstraint: core.UniqueConstraint<
      any,
      MariaDBConnector
    >,
  ) {
    assert(
      !uniqueConstraint.isIdentifier(),
      `The "${uniqueConstraint}" unique-constraint is the identifier`,
    );

    super(table);

    // config
    {
      this.config = uniqueConstraint.config.index;
      this.configPath = utils.addPath(uniqueConstraint.configPath, 'index');

      utils.assertNillablePlainObject(this.config, this.configPath);
    }
  }

  @Memoize()
  public override get columns(): ReadonlyArray<Column> {
    return Object.freeze(
      this.table.getColumnsByComponents(
        ...this.uniqueConstraint.componentsByName.values(),
      ),
    );
  }

  @Memoize()
  public override get name(): string {
    const nameConfig = this.config?.name;
    const nameConfigPath = utils.addPath(this.configPath, 'name');

    return nameConfig
      ? ensureIdentifierName(nameConfig, nameConfigPath)
      : this.table.schema.namingStrategy.getUniqueIndexName(this);
  }

  /**
   * @see https://mariadb.com/kb/en/create-table/#unique
   */
  @Memoize()
  public override get definition(): string {
    return `UNIQUE INDEX ${escapeIdentifier(this.name)} (${this.columns
      .map(({ name }) => escapeIdentifier(name))
      .join(',')})`;
  }
}
