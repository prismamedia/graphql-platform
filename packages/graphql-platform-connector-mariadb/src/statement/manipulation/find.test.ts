import {
  ArticleStatus,
  myAdminContext,
  MyGP,
  myUserContext,
} from '@prismamedia/graphql-platform/__tests__/config.js';
import { fixtures } from '@prismamedia/graphql-platform/__tests__/fixture.js';
import { format } from '@sqltools/formatter';
import { MariaDBConnector } from '../../index.js';
import { createGraphQLPlatform } from '../../__tests__/config.js';

describe('Find statement', () => {
  let gp: MyGP<MariaDBConnector>;

  beforeAll(async () => {
    gp = createGraphQLPlatform('connector_mariadb_find_statement');

    await gp.connector.setup();
    await gp.seed(fixtures, myAdminContext);
  });

  afterAll(async () => {
    await gp.connector.teardown();
  });

  it.each([
    [
      'Article',
      {
        orderBy: ['createdAt_ASC'],
        first: 5,
        selection: `{
          status
          title
          body
          views
          score
          machineTags
          metas
          tagCount
          createdAt
        }`,
      },
      myUserContext,
    ],
    [
      'Article',
      {
        orderBy: ['createdAt_ASC'],
        first: 5,
        selection: `{
          status
          title
          body
          views
          score
          machineTags
          metas
          tagCount
          createdAt
        }`,
      },
      myAdminContext,
    ],
    [
      'Article',
      {
        where: {
          OR: [
            { body_is_null: true },
            { body_contains: 'my +searched -text here' },
          ],
          status: ArticleStatus.PUBLISHED,
          category: { slug: 'tv' },
          createdAt_gte: '2022-01-01T00:00:00Z',
          createdBy: {
            profile: {
              OR: [
                { facebookId_is_null: false },
                { googleId_is_null: false },
                { twitterHandle_is_null: false },
              ],
            },
          },
          tagCount_gt: 5,
          tags_some: { tag: { deprecated_not: true } },
        },
        orderBy: ['createdAt_ASC'],
        first: 5,
        selection: `{
          status
          title
          category {
            title
          }
          createdBy {
            username
            profile {
              facebookId
              googleId
              twitterHandle
            }
          }
          allTagCount: tagCount
          filteredTagCount: tagCount(where: { tag_not: { deprecated: true }})
          allTags: tags(orderBy: [order_ASC], first: 5) {
            order
            tag {
              title
              deprecated
            }
          }
          filteredTags: tags(where: { tag_not: { deprecated: true }}, orderBy: [order_ASC], first: 5) {
            order
            tag {
              title
            }
          }
        }`,
      },
      myAdminContext,
    ],
  ])('generates statements', async (nodeName, args, context) => {
    const statements: string[] = [];

    const subscriber = gp.connector.executedStatements.subscribe(
      ({ statement }) => {
        statements.push(format(statement.sql).replaceAll('<= >', '<=>'));
      },
    );

    try {
      await expect(
        gp
          .getNodeByName(nodeName)
          .getQueryByKey('find-many')
          .execute(args, context),
      ).resolves.toMatchSnapshot();
    } finally {
      subscriber.unsubscribe();
    }

    expect(statements).toMatchSnapshot();
  });
});
