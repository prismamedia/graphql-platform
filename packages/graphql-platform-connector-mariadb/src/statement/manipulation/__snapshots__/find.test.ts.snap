// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Find statement generates statements 1`] = `
[
  {
    "body": null,
    "createdAt": 2022-03-01T00:00:00.000Z,
    "machineTags": null,
    "metas": null,
    "score": 0.5,
    "status": "published",
    "tagCount": 2,
    "title": "My first published article",
    "views": 0n,
  },
  {
    "body": {
      "blocks": [],
      "entityMap": {},
    },
    "createdAt": 2022-04-01T00:00:00.000Z,
    "machineTags": [
      "namespace:key=a_value",
      "namespace:key=other_value",
    ],
    "metas": {
      "aKey": "withAnyValue",
    },
    "score": 0.12,
    "status": "published",
    "tagCount": 1,
    "title": "My second published article",
    "views": 1234567890n,
  },
]
`;

exports[`Find statement generates statements 2`] = `
[
  "SELECT JSON_OBJECT(
    "status",
    \`articles\`.\`status\`,
    "title",
    \`articles\`.\`title\`,
    "body",
    \`articles\`.\`body\`,
    "views",
    \`articles\`.\`views\`,
    "score",
    \`articles\`.\`score\`,
    "machineTags",
    \`articles\`.\`machine_tags\`,
    "metas",
    \`articles\`.\`metas\`,
    "tagCount",
    (
      SELECT COUNT(*)
      FROM \`article_tags\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
    ),
    "createdAt",
    \`articles\`.\`created_at\`
  ) as \`Article\`
FROM \`articles\`
WHERE \`articles\`.\`status\` <=> 'published'
ORDER BY \`articles\`.\`created_at\` ASC
LIMIT 5",
]
`;

exports[`Find statement generates statements 3`] = `
[
  {
    "body": {
      "blocks": [],
      "entityMap": {},
    },
    "createdAt": 2022-01-01T00:00:00.000Z,
    "machineTags": null,
    "metas": {
      "myArbitraryKey": "with my arbitrary value",
    },
    "score": 0.5,
    "status": "draft",
    "tagCount": 0,
    "title": "My first draft article",
    "views": 0n,
  },
  {
    "body": null,
    "createdAt": 2022-02-01T00:00:00.000Z,
    "machineTags": null,
    "metas": null,
    "score": 0.5,
    "status": "draft",
    "tagCount": 0,
    "title": "My second draft article",
    "views": 0n,
  },
  {
    "body": null,
    "createdAt": 2022-03-01T00:00:00.000Z,
    "machineTags": null,
    "metas": null,
    "score": 0.5,
    "status": "published",
    "tagCount": 2,
    "title": "My first published article",
    "views": 0n,
  },
  {
    "body": {
      "blocks": [],
      "entityMap": {},
    },
    "createdAt": 2022-04-01T00:00:00.000Z,
    "machineTags": [
      "namespace:key=a_value",
      "namespace:key=other_value",
    ],
    "metas": {
      "aKey": "withAnyValue",
    },
    "score": 0.12,
    "status": "published",
    "tagCount": 1,
    "title": "My second published article",
    "views": 1234567890n,
  },
]
`;

exports[`Find statement generates statements 4`] = `
[
  "SELECT JSON_OBJECT(
    "status",
    \`articles\`.\`status\`,
    "title",
    \`articles\`.\`title\`,
    "body",
    \`articles\`.\`body\`,
    "views",
    \`articles\`.\`views\`,
    "score",
    \`articles\`.\`score\`,
    "machineTags",
    \`articles\`.\`machine_tags\`,
    "metas",
    \`articles\`.\`metas\`,
    "tagCount",
    (
      SELECT COUNT(*)
      FROM \`article_tags\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
    ),
    "createdAt",
    \`articles\`.\`created_at\`
  ) as \`Article\`
FROM \`articles\`
ORDER BY \`articles\`.\`created_at\` ASC
LIMIT 5",
]
`;

exports[`Find statement generates statements 5`] = `[]`;

exports[`Find statement generates statements 6`] = `
[
  "SELECT JSON_OBJECT(
    "status",
    \`articles\`.\`status\`,
    "title",
    \`articles\`.\`title\`,
    "category",
    IF(
      \`articles>categories\`.\`privateId\` IS NOT NULL,
      JSON_OBJECT("title", \`articles>categories\`.\`title\`),
      NULL
    ),
    "createdBy",
    IF(
      \`articles>users\`.\`id\` IS NOT NULL,
      JSON_OBJECT(
        "username",
        \`articles>users\`.\`username\`,
        "profile",
        IF(
          \`articles>users>user_profiles\`.\`theUserId\` IS NOT NULL,
          JSON_OBJECT(
            "facebookId",
            \`articles>users>user_profiles\`.\`facebook_id\`,
            "googleId",
            \`articles>users>user_profiles\`.\`google_id\`,
            "twitterHandle",
            \`articles>users>user_profiles\`.\`twitter_handle\`
          ),
          NULL
        )
      ),
      NULL
    ),
    "allTagCount",
    (
      SELECT COUNT(*)
      FROM \`article_tags\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
    ),
    "filteredTagCount",
    (
      SELECT COUNT(*)
      FROM \`article_tags\`
        INNER JOIN \`tags\` AS \`article_tags>tags\` ON \`article_tags\`.\`tag_id\` <=> \`article_tags>tags\`.\`id\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
        AND NOT (
          \`article_tags>tags\`.\`id\` IS NOT NULL
          AND \`article_tags>tags\`.\`deprecated\` <=> 1
        )
    ),
    "allTags",
    (
      SELECT JSON_ARRAYAGG(
          JSON_OBJECT(
            "order",
            \`article_tags\`.\`order\`,
            "tag",
            IF(
              \`article_tags>tags\`.\`id\` IS NOT NULL,
              JSON_OBJECT(
                "title",
                \`article_tags>tags\`.\`title\`,
                "deprecated",
                \`article_tags>tags\`.\`deprecated\`
              ),
              NULL
            )
          )
          ORDER BY \`article_tags\`.\`order\` ASC
        )
      FROM \`article_tags\`
        INNER JOIN \`tags\` AS \`article_tags>tags\` ON \`article_tags\`.\`tag_id\` <=> \`article_tags>tags\`.\`id\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
      ORDER BY \`article_tags\`.\`order\` ASC
      LIMIT 5
    ), "filteredTags", (
      SELECT JSON_ARRAYAGG(
          JSON_OBJECT(
            "order",
            \`article_tags\`.\`order\`,
            "tag",
            IF(
              \`article_tags>tags\`.\`id\` IS NOT NULL,
              JSON_OBJECT("title", \`article_tags>tags\`.\`title\`),
              NULL
            )
          )
          ORDER BY \`article_tags\`.\`order\` ASC
        )
      FROM \`article_tags\`
        INNER JOIN \`tags\` AS \`article_tags>tags\` ON \`article_tags\`.\`tag_id\` <=> \`article_tags>tags\`.\`id\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
        AND NOT (
          \`article_tags>tags\`.\`id\` IS NOT NULL
          AND \`article_tags>tags\`.\`deprecated\` <=> 1
        )
      ORDER BY \`article_tags\`.\`order\` ASC
      LIMIT 5
    )
  ) as \`Article\`
FROM \`articles\`
  LEFT JOIN \`categories\` AS \`articles>categories\` ON \`articles\`.\`category_privateId\` <=> \`articles>categories\`.\`privateId\`
  LEFT JOIN \`users\` AS \`articles>users\` ON \`articles\`.\`created_by_id\` <=> \`articles>users\`.\`id\`
  LEFT JOIN \`user_profiles\` AS \`articles>users>user_profiles\` ON \`articles>users\`.\`id\` <=> \`articles>users>user_profiles\`.\`theUserId\`
WHERE (
    \`articles\`.\`status\` <=> 'published'
    AND (
      \`articles>categories\`.\`privateId\` IS NOT NULL
      AND \`articles>categories\`.\`slug\` <=> 'tv'
    )
    AND (
      \`articles>users\`.\`id\` IS NOT NULL
      AND (
        \`articles>users>user_profiles\`.\`theUserId\` IS NOT NULL
        AND (
          NOT \`articles>users>user_profiles\`.\`facebook_id\` <=> NULL
          OR NOT \`articles>users>user_profiles\`.\`google_id\` <=> NULL
          OR NOT \`articles>users>user_profiles\`.\`twitter_handle\` <=> NULL
        )
      )
    )
    AND \`articles\`.\`created_at\` >= '2022-01-01 00:00:00.000'
    AND EXISTS (
      SELECT *
      FROM \`article_tags\`
        INNER JOIN \`tags\` AS \`article_tags>tags\` ON \`article_tags\`.\`tag_id\` <=> \`article_tags>tags\`.\`id\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
        AND (
          \`article_tags>tags\`.\`id\` IS NOT NULL
          AND NOT \`article_tags>tags\`.\`deprecated\` <=> 1
        )
    )
    AND (
      SELECT COUNT(*)
      FROM \`article_tags\`
      WHERE \`articles\`.\`privateId\` <=> \`article_tags\`.\`article_privateId\`
    ) > 5
    AND (
      \`articles\`.\`body\` <=> NULL
      OR MATCH (\`articles\`.\`body\`) AGAINST ('my searched text here')
    )
  )
ORDER BY \`articles\`.\`created_at\` ASC
LIMIT 5",
]
`;
